crates/bevy_ecs/src/world/mod.rs

```diff
@@ -96,13 +97,44 @@ use unsafe_world_cell::{UnsafeEntityCell, UnsafeWorldCell};
 ///     commands.queue(AddToCounter(42));
 /// }
 /// ```
-pub trait Command: Send + 'static {
+pub trait Command<Marker = ()>: Send + 'static {
     /// Applies this command, causing it to mutate the provided `world`.
     ///
     /// This method is used to define what a command "does" when it is ultimately applied.
     /// Because this method takes `self`, you can store data or settings on the type that implements this trait.
     /// This data is set by the system or other source of the command, and then ultimately read in this method.
-    fn apply(self, world: &mut World);
+    fn apply(self, world: &mut World) -> Result;
+
+    /// Returns a new [`Command`] that, when applied, will apply the original command
+    /// and handle any resulting errors according to the provided [`CommandErrorMode`].
+    fn with_error_handling(
+        self,
+        error_mode: CommandErrorMode,
+    ) -> impl FnOnce(&mut World) + Send + 'static
+    where
+        Self: Sized,
+    {
+        move |world: &mut World| {
+            if let Err(error) = self.apply(world) {
+                match error_mode {
+                    CommandErrorMode::Silent => (),
+                    CommandErrorMode::Warn => warn!("Command returned an error: {error}"),
+                    CommandErrorMode::Panic => panic!("Command returned an error: {error}"),
+                }
+            }
+        }
+    }
+}
+
+/// How a [`Command`] should respond upon encountering an [`Error`](crate::result::Error).
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum CommandErrorMode {
```

Comment:
I think that we need to be able to provide a closure here as well, for actually responding to the error in bespoke ways. This is relatively niche, but an important requirement. Unfortunately, I think we need to add that in the initial PR, since it's the most complex technical requirement and I can't confidently say whether or not this approach will handle it just by looking.

Reply:
I got it working with `fn()`, doing it with `FnOnce()` would be more complicated. Is `fn()` enough or should I keep working on it?

Reply:
That's enough :)